### PR TITLE
Daffodil 2280 cleanup - removes backpointers and factory patterns no longer needed

### PR DESCRIPTION
**Pull request by mbeckerle**
Requesting code review on this finally.

Big change set that cleans up and ultimately removes the backpointers that caused all that trouble in the schema compiler.

### CODE REVIEW

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/BitOrderMixin.scala [hunk 1]
```diff
@@ -45,34 +45,36 @@ trait BitOrderMixin extends GrammarMixin with ByteOrderAnalysisMixin { self: Ter
...
-          case None => false // no prior that has a bit order we could be the same as
-          case Some(prior) => {
-            if (prior.defaultBitOrder =:= this.defaultBitOrder) true
-            else false
-          }
-        }
-      }
-    res
-  }
+  final protected lazy val isKnownSameBitOrder: Boolean = false // FIXME: must analyze prior terms, enclosing terms, etc.
```
Comment (mbeckerle) [inline daffodil-core/src/main/scala/org/apache/daffodil/grammar/BitOrderMixin.scala]: Missed this. Need to investigate if this disables an important optimization. This file was converted to stubs and needs to be reconsidered.

File: daffodil-cli/src/it/scala/org/apache/daffodil/saving/TestCLISaveParser.scala [hunk 2]
```diff
@@ -189,11 +189,13 @@ class TestCLISaveParser {
...
-      shell.expect(contains("[debug]"))
+      // Too much output
+      // shell.sendLine(String.format("%s -vvv save-parser -s %s -r matrix %s", Util.binPath, testSchemaFile, savedParserFile.getName()))
+      // shell.expect(contains("[debug]"))
 
-      shell.sendLine(String.format("%s -vvvv save-parser -s %s -r matrix %s", Util.binPath, testSchemaFile, savedParserFile.getName()))
-      shell.expect(contains("[oolagdebug]"))
+      // Too much output
+      // shell.sendLine(String.format("%s -vvvv save-parser -s %s -r matrix %s", Util.binPath, testSchemaFile, savedParserFile.getName()))
+      // shell.expect(contains("[oolagdebug]"))
```
Comment (stevedlawrence) [inline daffodil-cli/src/it/scala/org/apache/daffodil/saving/TestCLISaveParser.scala]: Any chance this could be split into a separate pull request? I think we have a bug for removing the excessive test output.

File: daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ChoiceAndOtherVariousUnparsers.scala [hunk 3]
```diff
@@ -17,26 +17,59 @@
...
+import org.apache.daffodil.exceptions.Assert
+import org.apache.daffodil.util.PreSerialization
+
+case class ChoiceBranchMap(
+  lookupTable: Map[ChoiceBranchEvent, Unparser],
+  unmappedDefault: Option[Unparser])
+  extends PreSerialization {
+
+  @throws(classOf[java.io.IOException])
+  private def writeObject(out: java.io.ObjectOutputStream): Unit = serializeObject(out)
```
Comment (stevedlawrence) [inline daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ChoiceAndOtherVariousUnparsers.scala]: Why do we need custom serialization here?

File: daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ChoiceAndOtherVariousUnparsers.scala [hunk 4]
```diff
@@ -17,26 +17,59 @@
...
+        //
+        if (unmappedDefault.isDefined)
+          One(unmappedDefault.get)
+        else
+          Nope
+      }
+    res
+  }
+
+  def childProcessors = lookupTable.map { case (k, v) => v }.toSeq ++ unmappedDefault
```
Comment (stevedlawrence) [inline daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ChoiceAndOtherVariousUnparsers.scala]: Instead of the map, I think we can do loopupTable.values.toSeq, a little easier to read.

File: daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ChoiceAndOtherVariousUnparsers.scala [hunk 5]
```diff
@@ -53,11 +86,14 @@ class ChoiceCombinatorUnparser(
...
-    val childUnparser = eventUnparserMap.get(key).getOrElse {
+    val maybeChildUnparser = eventUnparserMap.get(key)
+    val childUnparser = if (maybeChildUnparser.isEmpty) {
       UnparseError(One(mgrd.schemaFileLocation), One(state.currentLocation),
         "Found next element %s, but expected one of %s.",
         key.qname.toExtendedSyntax,
         eventUnparserMap.keys.map { _.qname.toExtendedSyntax }.mkString(", "))
+    } else {
+      maybeChildUnparser.get
     }
```
Comment (stevedlawrence) [inline daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ChoiceAndOtherVariousUnparsers.scala]: Maybe easier to read if we do
```scala
if (maybeChildUnpaser.isEmpty) {
  UnparseErrror(...)
}
val childUnparser = maybeChildUnparser.get
```

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala [hunk 6]
```diff
@@ -175,7 +177,23 @@ class DataProcessor(
...
+    val sdp = new SerializableDataProcessor(ssrd, tunablesObj)
+    try {
+      //
+      // The serialization system encapsulates exceptions or any throw really
+      // inside an IOException.
+      //
+      // We catch and undo this, and throw the original exception,
+      // which won't give us a breakpoint at the right place, but at least
+      // it will give us the right exception, and if it is an SDE, a backtrace.
+      //
```
Comment (stevedlawrence) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DataProcessor.scala]: Does this comment mean that we could throw an SDE during serialization? Is this because of lazy evaluation? Doesn't the preserializations stuff prevent this?

Does this also mean that it's possible that an SDE that happens at compile time might not be checked and thrown until runtime?

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala [hunk 7]
```diff
@@ -195,8 +194,7 @@ sealed class NonTermRuntimeData(
   @TransientParam pathArg: => String,
   @TransientParam namespacesArg: => NamespaceBinding,
   @TransientParam unqualifiedPathStepPolicyArg: => UnqualifiedPathStepPolicy)
-  extends RuntimeData
-  with PreSerialization {
```
Comment (stevedlawrence) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala]: Why remove PreSerialization?

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ChoiceGroup.scala [hunk 8]
```diff
@@ -110,9 +110,12 @@ abstract class ChoiceTermBase(
...
 
-  requiredEvaluations(branchesAreNonOptional)
-  requiredEvaluations(branchesAreNotIVCElements)
-  requiredEvaluations(modelGroupRuntimeData.preSerialization)
+  override protected def check() {
+    super.check()
+    requiredEvaluations(branchesAreNonOptional)
+    requiredEvaluations(branchesAreNotIVCElements)
+    requiredEvaluations(modelGroupRuntimeData.preSerialization)
+  }
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/ChoiceGroup.scala]: This will need some scaladoc and a design note. 
A significant' change here is that in the past there were these factories. They did little except postpone the creation of the "real objects" until they were known participants in the schema. 

Upon  creation, the requiredEvaluation functions are registered, and *will* be evaluated later.

Now, we allow creation of the objects, and those objects which are always part of the "active schema"  like the schema set and processor factory object always can issue requiredEvaluations. But many other objects are created but we don't want to register their requiredEvaluations until it is known they are "in use by" the schema. Those have their calls to requiredEvaluations enclosed in a call to check(). Finally, a third class of things are only created on demand when another object, that has been determined to be "in use" causes them to be constructed. Those also require no qualification around their calls to requiredEvaluations(...) once constructed it is known they are used.

File: daffodil-cli/src/it/scala/org/apache/daffodil/saving/TestCLISaveParser.scala (hunk 2 above)
Comment (mbeckerle) [reply daffodil-cli/src/it/scala/org/apache/daffodil/saving/TestCLISaveParser.scala]: I can do that.

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ChoiceGroup.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/ChoiceGroup.scala]: Another comment addresses this also. It's because objects only want these requiredEcaluations to actually evaluate if the objects are "in use" by the schema. "in use" now isn't just "got constructed".

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ElementBase.scala [hunk 9]
```diff
@@ -288,7 +291,10 @@ trait ElementBase
     }
   }
 
-  private lazy val parentMinimizedScope = enclosingElement.map { _.minimizedScope }.getOrElse(scala.xml.TopScope)
+  private lazy val parentMinimizedScope = {
+    val ee = enclosingElements.headOption // FIXME: doesn't work unless all are same.
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/ElementBase.scala]: Ah. I have not dealt with the minimized scope issue as yet. That remains a TODO/FIXME issue here. 
This is assuming the right answer for parentMinimizedScope is that of the first shared instance. Because we only share things that share properties (shared PropEnv), this works in most cases. Not clear it will for this scope minimizaiton stuff because that's not really about properties.  So this will need fixing.

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ChoiceGroup.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/ChoiceGroup.scala]: Makes a lot of sense. Definitely needs some documntation. Thoughts on renaming the "check" function to something like ``registerAsActive`` to make it more clear what the purpose of it?

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ElementBase.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/ElementBase.scala]: Gotcha. Not worth fixing as part of this cleanup. Maybe just include the bug number so it's clear what this fixup references?

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ElementBase.scala [hunk 10]
```diff
@@ -502,41 +495,58 @@ trait ElementBase
...
+            case AlignmentUnits.Bytes => 8 * align
+          }
+          alignInBits
+        }
       }
-      case align: JInt => {
-        val alignInBits: JInt = this.alignmentUnits match {
-          case AlignmentUnits.Bits => align
-          case AlignmentUnits.Bytes => 8 * align
+    if ((alignment ne AlignmentType.Implicit) && this.isSimpleType) {
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/ElementBase.scala]: This adds the checking for an obscure case where an aligned thing also has an initiator where the text alignment of the charset may be incompatible with the term's own specified alignment.

[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/ElementBase.scala]: Comments being added to clarify.

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/Nesting.scala [hunk 11]
```diff
@@ -28,117 +29,80 @@ import org.apache.daffodil.exceptions.Assert
...
+                  case _ => Assert.invariantFailed("non-global component has schema document as lexical parent: " + this)
+                }
+              }
+              case ec => {
+                Seq(EnclosingComponentDef(ec, pos))
+              }
+            }
+          ecs
+      }
+    ecs
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/Nesting.scala]: Adding comments.

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/SchemaComponent.scala [hunk 12]
```diff
@@ -64,11 +64,25 @@ trait SchemaComponent
 
   override def oolagContextViaArgs = optLexicalParent
 
+  /**
+   * Override and call super.check() to indicate
+   * things that need to be checked if the object is being used, that is.
+   */
+  protected def check(): Unit = {
```
[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/SchemaComponent.scala]: I like it!  Maybe adds a little extra complication, but makes the difference more intuitive.

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ParticleMixin.scala [hunk 13]
```diff
@@ -96,7 +96,11 @@ trait RequiredOptionalMixin { self: ElementBase =>
       // Treat all arrays as non-required so that we can tolerate invalid
       // infosets where the number of events is fewer than the array minimum occurrences.
       else if (isArray) false
-      else false
+      else if (minOccurs == 0 && maxOccurs == 0) false
```
[the remark this answers is not in the document]
Comment (stevedlawrence) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/ParticleMixin.scala]: Yeah, according to the spec (https://www.w3.org/TR/2012/REC-xmlschema11-1-20120405/structures.html#declare-element) maxOccurs == 0 is allowed--it has to be a nonNegativeInteger or "unbounded".

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala (hunk 7 above)
Comment (mbeckerle) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/RuntimeData.scala]: Moved to base trait. I needed RuntimeData to be serializable.

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ModelGroup.scala [hunk 14]
```diff
@@ -417,7 +370,7 @@ abstract class ModelGroup(index: Int)
    */
   final def mustHaveRequiredElement: Boolean = LV('mustHaveRequiredElement) {
     this match {
-      case s: SequenceTermBase if s.isHidden => false
+      case s: SequenceTermBase if s.isHiddenGroupRef => false
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/ModelGroup.scala]: Yes. This case analysis really shouldn't use SequenceTermBase. I think you removed this member and added it to only the SGR/CGR class perhaps. That will come out in the merge.

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ComplexTypes.scala [hunk 15]
```diff
@@ -84,29 +87,16 @@ sealed abstract class ComplexTypeBase(xmlArg: Node, parentArg: SchemaComponent)
...
-  val elementDecl: ElementDeclMixin,
-  override val factory: GlobalComplexTypeDefFactory)
+  schemaDocumentArg: SchemaDocument)
   extends ComplexTypeBase(xmlArg, schemaDocumentArg)
   with GlobalNonElementComponentMixin
-  with NestingTraversesToReferenceMixin {
-
-  override lazy val referringComponent = Option(elementDecl)
+  with NestingLexicalMixin {
+  // nothing here
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/ComplexTypes.scala]: will do. It's because the base classes/traits are providing all the necessary functionality. There is a local mixin in one class, and a global mixin in the other, but otherwise these two classes are pretty much the same, and the base class provides everything.

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/GroupRef.scala [hunk 16]
```diff
@@ -53,24 +53,43 @@ trait GroupRef { self: ModelGroup =>
...
+    f.groupRef
+  }
+}
+
+/**
+ * This is a quasi schema component. It is such simply to share a bunch of
+ * mechanism that schema components inherit, in a hassle free manner.
+ * (E.g., diagnostics mechanism that keeps track of file and line number).
+ */
+final class GroupRefFactory private (refXML: Node, val refLexicalParent: SchemaComponent, position: Int, isHidden: Boolean)
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/GroupRef.scala]: For a class I think you have to put the private before the "(" of the arglist because this makes the constructor private, not the class. I only want the constructor method to be private which forces use of the companion object to construct the instances because the companion object is allowed to see private methods/members of the class.

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/GroupRef.scala [hunk 17]
```diff
@@ -53,24 +53,43 @@ trait GroupRef { self: ModelGroup =>
...
+object GroupRefFactory {
+
+  def apply(refXML: Node, refLexicalParent: SchemaComponent, position: Int, isHidden: Boolean) = {
+    val f = new GroupRefFactory(refXML, refLexicalParent, position, isHidden)
+    f.groupRef
+  }
+}
+
+/**
+ * This is a quasi schema component. It is such simply to share a bunch of
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/GroupRef.scala]: Improved comment language.

File: daffodil-lib/src/main/scala/org/apache/daffodil/api/Diagnostic.scala [hunk 18]
```diff
@@ -21,29 +21,64 @@ import org.apache.daffodil.util.Misc
...
+  Assert.invariant(maybeCause.isDefined || maybeFormatString.isDefined)
+}
+/**
+ * Base class for all thick or thin error, warning, info, and other sorts of objects
+ * that capture diagnostic information. Such as Schema Definition Errors/Warnings.
  *
  * Allows for lazy message creation, internationalization, etc.
  */
-abstract class Diagnostic(
+abstract class Diagnostic protected (
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-lib/src/main/scala/org/apache/daffodil/api/Diagnostic.scala]: Again this is protected constructor. This idiom is so that this particular constructor, the primary one with all the various confusing parameters, is only available for derived classes or more tidy public constructors to use.  

As for what this actually means. This means derived classes will implement a constructor with this same big all-the-args signature, and that one will ahve this same visibility protected. Their public constructors are the ones in the API intended for use.

File: daffodil-core/src/main/scala/org/apache/daffodil/dsom/ModelGroup.scala [hunk 19]
```diff
@@ -357,7 +349,7 @@ abstract class ModelGroup(index: Int)
             // follow it
             Seq(e) ++ e.possibleNextSiblingTerms
           }
-          case Some(s: SequenceTermBase) if s.isHidden => s.possibleNextSiblingTerms
+          case Some(s: SequenceTermBase) if s.isHiddenGroupRef => s.possibleNextSiblingTerms
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/dsom/ModelGroup.scala]: Yes. Yes.

File: daffodil-core/src/main/scala/org/apache/daffodil/runtime1/TermRuntime1Mixin.scala [hunk 20]
```diff
@@ -101,6 +110,91 @@ trait TermRuntime1Mixin { self: Term =>
...
+      // may also be terminated by finding the next element after the array,
+      // so we get closed only if required and not an array.
+      //
+      case eb: ElementBase if (eb.isRequiredStreamingUnparserEvent) =>
+        Closed(Seq(PNE(eb, true)))
+      case eb: ElementBase =>
+        Open(Seq(PNE(eb, false)))
+      case stb: SequenceTermBase if stb.isHiddenGroupRef =>
+        Open(Nil)
+      case stb: ChoiceTermBase if stb.isHiddenGroupRef =>
```
[the remark this answers is not in the document]
Comment (mbeckerle) [reply daffodil-core/src/main/scala/org/apache/daffodil/runtime1/TermRuntime1Mixin.scala]: check!
